File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let s = 1;
    let e = 100000+ 1;
    while (e > s) {
        let chk_level = Math.floor((e + s) / 2);
        let chk_time = 0;
        
        for (let i = 0; i < diffs.length; i++) {
            let time_cur = times[i];
            let time_pre = 0
            if (i !== 0) {
                time_pre = times[i - 1]
            }
            let used_time = time_cur + Math.max(0, (diffs[i] - chk_level)) * (time_cur + time_pre);
            chk_time += used_time;
        }

        if (chk_time <= limit) {
            e = chk_level;
        } else {
            s = chk_level + 1;
        }
    }
    return s;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    s,e = 1, max(diffs)+1
    
    while e > s:
        chk_level = (e+s) // 2
        chk_time = 0
        
        for i in range(len(diffs)):
            if i == 0:
                time_cur=times[0]
                time_pre=0
            else:
                time_cur=times[i]
                time_pre=times[i-1]
            used_time = time_cur + max(0,(diffs[i]-chk_level))*(time_cur+time_pre)
            chk_time += used_time
            
        if chk_time <= limit:
            e=chk_level
        else:
            s=chk_level+1
    return s

    